### ISSUE TITLE
[Phase 1] Data Pipeline and Feature Extraction

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/predictive-maintenance-survival#2]
## Summary

Implement the data loading and feature extraction pipeline for NASA C-MAPSS turbofan engine degradation dataset. This phase establishes the foundation for all subsequent survival modeling work.

**Parent Issue:** #1

## Objectives

- Load and preprocess C-MAPSS dataset (all 4 subsets: FD001-FD004)
- Extract meaningful statistical features from sensor time series
- Format data for survival analysis (duration, event indicator, covariates)

## System Context

```
data/raw/CMAPSSData/
├── train_FD001.txt    # 100 engines, single operating condition
├── test_FD001.txt     # 100 engines (cut before failure)
├── RUL_FD001.txt      # True RUL for test engines
├── train_FD002.txt    # 259 engines, 6 operating conditions
├── ...                # Similar for FD003, FD004
```

## Files to Create

| File | Description |
|------|-------------|
| `src/data/loader.py` | C-MAPSS data loading with RUL computation |
| `src/data/preprocessor.py` | Normalization, operating condition handling |
| `src/features/time_domain.py` | Rolling statistics, trend features |
| `src/features/extractor.py` | Main feature extraction pipeline |
| `tests/test_data_loader.py` | Unit tests for data loading |
| `tests/test_feature_extraction.py` | Unit tests for features |

### Data Loader
- [ ] Implement `load_dataset()` function for all 4 subsets
- [ ] Implement `compute_rul()` for training data
- [ ] Implement `load_rul_targets()` for test ground truth
- [ ] Handle whitespace-separated format correctly

### Preprocessor
- [ ] Per-sensor normalization (z-score)
- [ ] Operating condition grouping for FD002/FD004
- [ ] Train/test split by engine ID (no leakage)

### Feature Extraction
- [ ] Rolling mean, std, min, max (windows: 5, 10, 20, 50)
- [ ] Skewness, kurtosis for each sensor
- [ ] Linear trend slope and R-squared
- [ ] Sensor correlation features

## Code Skeleton

📄 **File:** `src/data/loader.py`

```python
"""NASA C-MAPSS dataset loader for turbofan engine degradation data."""

from pathlib import Path
import pandas as pd
import numpy as np
from typing import Tuple

COLUMN_NAMES = (
    ['engine_id', 'cycle'] +
    ['setting_1', 'setting_2', 'setting_3'] +
    [f's_{i}' for i in range(1, 22)]
)


def load_dataset(data_dir: Path, subset: str = 'FD001') -> Tuple[pd.DataFrame, pd.DataFrame]:
    """Load NASA C-MAPSS turbofan engine degradation dataset.

    Args:
        data_dir: Path to directory containing the txt data files
        subset: One of 'FD001', 'FD002', 'FD003', 'FD004'

    Returns:
        Tuple of (train_df, test_df)
    """
    data_dir = Path(data_dir)

    valid_subsets = ['FD001', 'FD002', 'FD003', 'FD004']
    if subset not in valid_subsets:
        raise ValueError(f"subset must be one of {valid_subsets}")

    train_path = data_dir / f'train_{subset}.txt'
    train_df = pd.read_csv(train_path, sep=r'\s+', header=None, names=COLUMN_NAMES)

    test_path = data_dir / f'test_{subset}.txt'
    test_df = pd.read_csv(test_path, sep=r'\s+', header=None, names=COLUMN_NAMES)

    return train_df, test_df


def compute_rul(df: pd.DataFrame) -> pd.DataFrame:
    """Add RUL column: max_cycle - current_cycle for each engine."""
    df = df.copy()
    max_cycles = df.groupby('engine_id')['cycle'].max()
    df['max_cycle'] = df['engine_id'].map(max_cycles)
    df['rul'] = df['max_cycle'] - df['cycle']
    return df.drop(columns=['max_cycle'])
```

📄 **File:** `src/features/time_domain.py`

```python
"""Time-domain statistical feature extraction."""

import numpy as np
from scipy import stats


def compute_rolling_statistics(
    series: np.ndarray,
    window_sizes: list[int] = [5, 10, 20, 50]
) -> dict[str, float]:
    """Compute rolling statistics for a sensor time series.

    Uses the last N values for each window size.
    """
    features = {}

    for w in window_sizes:
        if len(series) >= w:
            window = series[-w:]
            features[f'rolling_mean_{w}'] = np.mean(window)
            features[f'rolling_std_{w}'] = np.std(window)
            features[f'rolling_min_{w}'] = np.min(window)
            features[f'rolling_max_{w}'] = np.max(window)
        else:
            # Handle short series
            features[f'rolling_mean_{w}'] = np.mean(series)
            features[f'rolling_std_{w}'] = np.std(series)
            features[f'rolling_min_{w}'] = np.min(series)
            features[f'rolling_max_{w}'] = np.max(series)

    return features


def compute_trend_features(series: np.ndarray) -> dict[str, float]:
    """Compute trend-based degradation features."""
    x = np.arange(len(series))

    # Linear regression
    slope, intercept, r_value, p_value, std_err = stats.linregress(x, series)

    return {
        'trend_slope': slope,
        'trend_r_squared': r_value ** 2,
        'trend_intercept': intercept
    }
```

## Definition of Done

- [ ] `load_dataset()` returns correct shapes for all 4 subsets
- [ ] RUL values are non-negative and decrease monotonically per engine
- [ ] Feature extraction produces 50+ features without NaN values
- [ ] All tests pass: `pytest tests/test_data_loader.py tests/test_feature_extraction.py -v`
- [ ] Processing time < 30 seconds for FD001

## Risk Assessment

| Risk | Impact | Mitigation |
|------|--------|------------|
| File path issues on different OS | 🟡 | Use `pathlib.Path` consistently |
| Memory issues with large feature matrix | 🟢 | FD001 is small (~20K rows) |
| NaN in rolling features for short series | 🟡 | Handle edge cases explicitly |

## First Task

Create `src/data/loader.py` with the `load_dataset()` function. Test manually with:

```bash
python -c "from src.data.loader import load_dataset; print(load_dataset('data/raw/CMAPSSData')[0].shape)"
```
